collaboration-concept.md

```diff
@@ -0,0 +1,151 @@
+    3. Chat functionality
+    4. Output of host terminal
+2. Proxy message to host (with result or without):
+    1. Debug launch (notification)
+    2. File System information
+    3. Monaco language services (i.e. LSP requests)
+
+The following diagram shows the Session creation/entry negotiations. In all cases where the User/Host communicate directly in the diagram, the intermediary server provides the messaging exchange infrastructure (using a hole-punching mechanism). The host is always the user that provides the workspace
+
+```mermaid
+sequenceDiagram
+    participant Host
+    participant I as Intermediary
+    participant User
+    Host->>I: Create Collab Session
+    I->>Host: Return with Session ID
+    Host-->>User: Sends Session ID via chat
+    User->>Host: Request Session Entry
+    Host->>User: Allow Session Entry (with permissions)
+    User->>Host: Download Workspace Info
+```
+
+Diagram outlining the broadcasting mechanism. We skip the Intermediary here, since it only proxies messages between users. Also note that we don’t need a host, since in this message scenario, all users have “equal rights”. This example uses the notification for file changes, but can be used for broadcasts for debugging or terminal output
+
+```mermaid
+sequenceDiagram
+    participant U1 as User A
+    participant U2 as User B
+    participant U3 as User C
+  
+    Note over U1: Modifies a file
+	
+    par U1 to U2
+        U1-)U2: Send file notification event
+    and U1 to U2
+        U1-)U3: Send file notification event
+    end
+    par 
+        Note over U2: Updates modified file
+        Note over U3: Updates modified file
+    end 
+```
+
+Proxy host messaging example:
+
+```mermaid
+sequenceDiagram
+    participant P as Plugin Host
+    participant Host as Host User
+    participant User
+    Note over User: Opens File
+    Note over User: Performs Completion
+    User->>Host: Requests Completion
+    Host->>P: Request Completion
+    P->>Host: Return Completion
+    Host->>User: Return Completion
+    Note over User: Show Completion
+```
+
+## Coordination/Messaging Server
+
+For the messages architecture above, we need a coordination or messaging server. The server serves exactly two purposes:
+
+1. Opening up the host system to the world. The host and the coordination server negotiate a new collab session. This can be as simple as an HTTP call to this server. The coordination server then creates a virtual room and assigns the host that originated the call as its host.
+2. Communication between collaborators within a session. For example, a broadcast notification needs to be send to the coordination server first, which then relays the message to all other participants.
+
+Assuming the server is built specifically for Theia, we can reuse some of Theia’s dependencies for this, i.e. socket-io for bidirectional, high-performance messaging on top of HTTP.
+
+
+## API Concept
```

Comment:
What kind of API are we talking about here? Theia Extension API? VS Code Extension API?

Should we inject the `vsls` extension the same way we do with the `vscode` API so that extensions relying on this package can work in Theia?

Should we take care to properly implement the groundwork feature in the framework before exposing it through the plugin API? (By contrast with webviews which are baked-in the `plugin-ext` package at the moment.)

Reply:
Then is it fair to assume that this library must support at least what `vsls` supports? That would make for a good starting point for designing the API.